keps/sig-node/4112-passdown-resources-to-cri/README.md

```diff
@@ -0,0 +1,958 @@
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The kubelet does not provide complete information about the container
+resources specification (requests and limits) to CRI. However, various use
+cases have been identified where detailed knowledge of all the resources can be
+utilized in container runtimes for more optimal resource allocation to improve
+application performance and reduce cross-application interference.
+
+This KEP proposes a CRI API extension for disclosing container resource
+requests and limits to container runtimes.
+
+## Motivation
+
+Kubelet manages the native resources (CPU and memory) and communicates
+resource parameters over the CRI API to the runtime. However, the original
+details of the resource spec are lost as they get translated (within kubelet)
+to platform-specific (i.e. Linux or Windows) resource controller parameters
+like cpu shares, memory limits etc. Non-native resources such as extended
+resources and the device plugin resources are not visible to container runtimes
+at all.
+
+However, VM-based runtimes such as
+[Kata containers](https://katacontainers.io/), platform-optimized container
+runtimes,
+[OCI hooks](https://github.com/opencontainers/runtime-spec/blob/master/config.md),
+[runC](https://github.com/opencontainers/runc) wrappers,
+[NRI](https://github.com/containerd/nri) plugins or in some cases even
+applications themselves would benefit on getting full resource information,
+e.g. for reserving all required resources at pod sandbox creation (as it might
+be hard to impossible after that) or doing customized resource optimization.
+Extending the CRI API in to include the resource information would provide a
+comprehensive view of all resource usage of containers, allowing improved
+resource allocation without breaking any existing use cases.
+
+### Goals
+
+- make container resource spec transparently visible to CRI (the container
+  runtime)
+
+### Non-Goals
+
+- change kubelet resource management
+- change existing behavior of CRI
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a VM-based container runtime developer, I want to allocate/expose enough
+RAM, hugepages, GPU memory, protected memory sections and other resources for
+the VM to ensure that all containers in the pod are guaranteed to get the
+resources they require.
+
+#### Story 2
+
+As a platform-optimized CRI runtime developer, I want to know detailed
+container resource requests to be able to make optimal, platform specific,
+resource allocations. Some of the resources may be handled outside cgroups and
+be container runtime specific details.
+
+#### Story 3
+
+As a cluster administrator, I want to install an OCI hook/runc wrapper/NRI
```

Comment:
Could you expand on this use case? How does extending the CRI translate to modifications in the OCI runtime specification which is interpreted by runc (or wrappers)?

Reply:
The CRI changes (in this KEP) would not directly translate to anything in the OCI config. It's just "informational" that a possible hook/wrapper/plugin can then use to tweak the OCI config. Say you want to do customized cpu pinning in your plugin. I'll come up with some more flesh on this section...

Reply:
Thanks.

Reply:
Resolved?

Reply:
ping @elezar may I close this as resolved?